### PR TITLE
data source logging endpoints

### PR DESCRIPTION
**Pull request by EleanorKilgallon99**

### CODE REVIEW

File: examples/provider-verification/main.tf [hunk 1]
```diff
@@ -7,7 +7,13 @@ terraform {
 }
 
 provider "altitude" {
-  client_id     = "{{CLIENT ID}}"
-  client_secret = "{{CLIENT SECRET}}"
-  mode          = "Local"
+  client_id     = "sJ8kVG1yCFyD8qOTLAS5yEy2F28ZM7qO"
+  client_secret = "w455Q1EFsVkOKxm4u05Reuiy84mp1Ka3ihHSfHBctHvxNtzZ61W-HWxxETHzsfZN"
```
Comment (g-sreshtha) [inline examples/provider-verification/main.tf]: Is there a way to pass the id and secret down without hardcoding it?

Comment (EleanorKilgallon99) [reply examples/provider-verification/main.tf]: yeah thats not meant to be there - will delete this client

File: internal/provider/data_logging_endpoints.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ datasource.DataSource              = &LoggingEndpointsDataSource{}
+	_ datasource.DataSourceWithConfigure = &LoggingEndpointsDataSource{}
+)
+
+// NewLoggingEndpointsDataSource is a helper function to simplify the provider implementation.
+func NewLoggingEndpointsDataSource() datasource.DataSource {
+	return &LoggingEndpointsDataSource{}
```
Comment (JosephMcDonagh) [inline internal/provider/data_logging_endpoints.go:21]: What does this do?

File: internal/provider/data_logging_endpoints.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+// coffeesDataSource is the data source implementation.
```
Comment (JosephMcDonagh) [inline internal/provider/data_logging_endpoints.go]: typo

File: internal/provider/data_logging_endpoints.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+											},
+											"defaultvalue": schema.StringAttribute{
+												Computed: true,
+											},
+										},
+									},
+								},
+								"secretkey": schema.StringAttribute{
+									Computed: true,
+									Optional: true,
```
Comment (JosephMcDonagh) [inline internal/provider/data_logging_endpoints.go]: Why is secret key optional?

File: internal/provider/data_logging_endpoints.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+	for _, endpoint := range loggingEndpoints.Endpoints {
+		endpointState := LoggingEndpointDataSourceModel{
+			Type:          types.StringValue(endpoint.Type),
+			EnvironmentId: types.StringValue(endpoint.EnvironmentId),
+			Config:        transformToConfigResourceModel(endpoint.Config),
+			ID:            types.StringValue("placeholder"),
+		}
+
+		state.Endpoints = append(state.Endpoints, endpointState)
```
Comment (JosephMcDonagh) [inline internal/provider/data_logging_endpoints.go:154]: Can this use the same format as `headerModels[i] = headersBody` with the `i` for consistency. Or the one below use append. ([i] might be quicker than append though, not sure)

File: internal/provider/data_logging_endpoints.go (hunk 2 above)
Comment (EleanorKilgallon99) [reply internal/provider/data_logging_endpoints.go:21]: According to the docs its a helper function to simplify the provider implementation.

File: internal/provider/data_logging_endpoints.go (hunk 5 above)
Comment (EleanorKilgallon99) [reply internal/provider/data_logging_endpoints.go:160]: done

File: internal/provider/data_logging_endpoints.go [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+// loggingEndpointsDataSourceModel maps the data source schema data.
+
+type LoggingEndpointsDataSourceModel struct {
+	Endpoints []LoggingEndpointDataSourceModel `tfsdk:"endpoints"`
+}
+type LoggingEndpointDataSourceModel struct {
+	ID            types.String                        `tfsdk:"id"`
+	Type          types.String                        `tfsdk:"type"`
+	EnvironmentId types.String                        `tfsdk:"environmentid"`
```
Comment (OLingard) [inline internal/provider/data_logging_endpoints.go]:
```suggestion
	EnvironmentId types.String                        `tfsdk:"environment_id"`
```

File: internal/provider/data_logging_endpoints.go [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+	Config        GetAbstractAccessLoggingConfigModel `tfsdk:"config"`
+}
+
+type GetAbstractAccessLoggingConfigModel struct {
+	Dataset   types.String           `tfsdk:"dataset"`
+	ProjectId types.String           `tfsdk:"projectid"`
```
Comment (OLingard) [inline internal/provider/data_logging_endpoints.go]:
```suggestion
	ProjectId types.String           `tfsdk:"project_id"`
```

File: internal/provider/data_logging_endpoints.go [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+	Email     types.String           `tfsdk:"email"`
+	Headers   []BqLoggingHeaderModel `tfsdk:"headers"`
+	SecretKey types.String           `tfsdk:"secretkey"`
+}
+
+type BqLoggingHeaderModel struct {
+	ColumnName   types.String `tfsdk:"columnname"`
+	HeaderName   types.String `tfsdk:"headername"`
+	DefaultValue types.String `tfsdk:"defaultvalue"`
+}
```
Comment (OLingard) [inline internal/provider/data_logging_endpoints.go]:
```suggestion
	SecretKey types.String           `tfsdk:"secret_key"`
}

type BqLoggingHeaderModel struct {
	ColumnName   types.String `tfsdk:"column_name"`
	HeaderName   types.String `tfsdk:"header_name"`
	DefaultValue types.String `tfsdk:"default_value"`
}
```

File: internal/provider/data_logging_endpoints.go [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+	for i, endpoint := range loggingEndpoints.Endpoints {
+		endpointState := LoggingEndpointDataSourceModel{
+			Type:          types.StringValue(endpoint.Type),
+			EnvironmentId: types.StringValue(endpoint.EnvironmentId),
+			Config:        transformToConfigResourceModel(endpoint.Config),
+			ID:            types.StringValue("placeholder"),
+		}
+
+		state.Endpoints[i] = endpointState
+	}
```
Comment (OLingard) [inline internal/provider/data_logging_endpoints.go]: Why is this replicated? And also why is the string value placeholder?

File: internal/provider/data_logging_endpoints.go [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+	Endpoints []LoggingEndpointDataSourceModel `tfsdk:"endpoints"`
+}
+type LoggingEndpointDataSourceModel struct {
+	ID            types.String                        `tfsdk:"id"`
+	Type          types.String                        `tfsdk:"type"`
+	EnvironmentId types.String                        `tfsdk:"environmentid"`
+	Config        GetAbstractAccessLoggingConfigModel `tfsdk:"config"`
+}
+
+type GetAbstractAccessLoggingConfigModel struct {
```
Comment (OLingard) [inline internal/provider/data_logging_endpoints.go]: Can we call this BQAccessLogging, it's not abstract. We can abstract it when it's required

File: internal/provider/data_logging_endpoints.go [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+									Computed: true,
+								},
+								"email": schema.StringAttribute{
+									Computed: true,
+								},
+								"headers": schema.ListNestedAttribute{
+									Computed: true,
+									NestedObject: schema.NestedAttributeObject{
+										Attributes: map[string]schema.Attribute{
+											"columnname": schema.StringAttribute{
```
Comment (OLingard) [inline internal/provider/data_logging_endpoints.go]: These also need changing to snake case

File: internal/provider/data_logging_endpoints.go (hunk 5 above)
Comment (EleanorKilgallon99) [reply internal/provider/data_logging_endpoints.go:154]: I disagree as the headerModels[i] = headersBody is replacing items in the list, the list already exists so append wouldn't apply here. Whereas for state.Endpoints, the list doesnt exist. So state.Endpoints[i] doesnt exist and an error will be thrown.

File: internal/provider/data_logging_endpoints_test.go [hunk 12]
```diff
@@ -0,0 +1,28 @@
+package provider
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+)
+
+func TestAccLoggingEndpointsDataSource(t *testing.T) {
```
Comment (OLingard) [inline internal/provider/data_logging_endpoints_test.go:9]: I think we should remove the Github test which tests this on Prod. It's risky having this test on Prod as if it fails we'd have leaked data. Please could you remove the Github workflow to run on Prod?

**Review by OLingard — APPROVED** [review]
LGTM

**Pull request merged**